### PR TITLE
Use name for output variables instead of out in JIT

### PR DESCRIPTION
**Pull request by cpuhrsch**
This adds 88 matches.

### CODE REVIEW

**Review by wanchaol — APPROVED** [review]

**Comment by zdevito** [comment]
I am pretty sure that renaming line was necessary to make the `out` keyword-only argument work. It was added in this PR: https://github.com/pytorch/pytorch/pull/13093/files which adds support for that. It is possible it is no longer necessary if all the `_out` variants have renamed their output variable to `out`.

**Comment by cpuhrsch** [comment]
@zdevito - ok, it looks like it's worth digging deeper into this.

**Comment by cpuhrsch** [comment]
The generated code [changeset](https://gist.github.com/cpuhrsch/fe13cce8610b6e86b377e6ff05a18572) affects

```
        modified:   build/aten/src/ATen/Declarations.yaml
        modified:   torch/csrc/jit/generated/register_aten_ops_0.cpp
        modified:   torch/csrc/jit/generated/register_aten_ops_1.cpp
        modified:   torch/csrc/jit/generated/register_aten_ops_2.cpp
```

but doesn't impact any Python functions. Since the JIT doesn't care about the names of keyword arguments, it looks like these functions aren't exposed via Python anyway. All other functions for which it matters are thus presumably already using `out` as the output name as is required by the Python frontend.

File: aten/src/ATen/native/native_functions.yaml [hunk 1]
```diff
@@ -4525,119 +4589,136 @@
...
+  matches_jit_signature: True
   python_module: nn
 
-- func: sigmoid_backward(Tensor grad_output, Tensor output) -> Tensor
+- func: sigmoid_backward(Tensor grad_output, Tensor out) -> Tensor
   matches_jit_signature: True
   python_module: nn
 
-- func: tanh_backward(Tensor grad_output, Tensor output, *, Tensor(a!) grad_input) -> Tensor(a!)
+- func: tanh_backward(Tensor grad_output, Tensor out, *, Tensor(a!) grad_input) -> Tensor(a!)
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml]: here too.

File: aten/src/ATen/native/native_functions.yaml [hunk 2]
```diff
@@ -4525,119 +4589,136 @@
...
 - func: upsample_nearest3d_backward(Tensor grad_output, int[3] output_size, int[5] input_size, *, Tensor(a!) grad_input) -> Tensor(a!)
+  matches_jit_signature: True
   python_module: nn
 
 - func: upsample_nearest3d_backward(Tensor grad_output, int[3] output_size, int[5] input_size) -> Tensor
   matches_jit_signature: True
   python_module: nn
 
-- func: sigmoid_backward(Tensor grad_output, Tensor output, *, Tensor(a!) grad_input) -> Tensor(a!)
+- func: sigmoid_backward(Tensor grad_output, Tensor out, *, Tensor(a!) grad_input) -> Tensor(a!)
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml]: this looks weird -- output is an output in this case.

File: aten/src/ATen/native/native_functions.yaml [hunk 3]
```diff
@@ -4525,119 +4589,136 @@
...
+  matches_jit_signature: True
   python_module: nn
 
-- func: sigmoid_backward(Tensor grad_output, Tensor output) -> Tensor
+- func: sigmoid_backward(Tensor grad_output, Tensor out) -> Tensor
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml]: here too.

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -4525,119 +4589,136 @@
...
   matches_jit_signature: True
   python_module: nn
 
-- func: tanh_backward(Tensor grad_output, Tensor output, *, Tensor(a!) grad_input) -> Tensor(a!)
+- func: tanh_backward(Tensor grad_output, Tensor out, *, Tensor(a!) grad_input) -> Tensor(a!)
+  matches_jit_signature: True
   python_module: nn
 
-- func: tanh_backward(Tensor grad_output, Tensor output) -> Tensor
+- func: tanh_backward(Tensor grad_output, Tensor out) -> Tensor
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml]: here too.

File: tools/jit/gen_jit_dispatch.py [hunk 5]
```diff
@@ -449,7 +449,7 @@ def match_signature(decl, constructed_string, should_match_schema):
 
 def signature(decl, should_match_schema=True):
     def format_arg(arg):
-        name = arg['name'] if not arg.get('output') else 'out'
+        name = arg['name']  # if not arg.get('output') else 'out'
```
Comment (gchanan) [inline tools/jit/gen_jit_dispatch.py]: can you clean this up?

File: tools/autograd/gen_python_functions.py [hunk 6]
```diff
@@ -891,6 +891,14 @@ def get_py_formal_arg(arg):
             typename = 'TensorList[{}]'.format(len(typenames))
         else:
             typename = typenames[0]
+        # The nn module bindings are not exposed to the user directly
+        # but via torch.nn modules and functionals.
+        if len(output_args) == 1 and declaration['python_module'] != 'nn':
+            assert output_args[0]['name'] == 'out', \
+                "For function named {}: Our Python frontend uses out as" \
+                " the keyword argument name for output arguments.".format(name)
+        # NB: For more than 1 output args the type name is a TensorList
```
Comment (gchanan) [inline tools/autograd/gen_python_functions.py]: is this note correct?  There are a bunch of cases with multiple output arguments that are Tensors AFAII.

Comment (cpuhrsch) [reply tools/autograd/gen_python_functions.py]: It's determined above. `typenames` is of the same length as `output_args` and `typename` is set to be `TensorList[len(typenames)]` if that's longer than 1.

**Comment by cpuhrsch** [comment]
[This](https://gist.github.com/cpuhrsch/2b7a2326070679b26a7e756088d7a1ba) is the codegen changeset.

File: aten/src/ATen/native/README.md [hunk 7]
```diff
@@ -112,6 +112,12 @@ argument name you provide, and a rename of an argument name is considered a BC-b
 change (e.g., you will probably need to update `tools/autograd/derivatives.yaml` at
 least). For more details please see the section on `variants`.
 
+As a convention we use 'out' to indicate an output argument. This aligns with the
```
Comment (gchanan) [inline aten/src/ATen/native/README.md]: nit: this message is a little strange.  "This aligns with the Python bindings" is true, but the python bindings should be generated based on the schema here (and I think most would expect it is).

This also doesn't apply to operators with multiple output arguments.

"Check the generated code when making a change
to make sure you're not breaking the API when renaming an argument name of an
existing function."

this applies to any parameter you rename, not just output arguments.

**Review by gchanan — APPROVED** [review]

**Pull request closed without merging**